In [1]:
# Importing the data csv file
import pandas as pd
veg_df = pd.read_csv("Excl_exp3.csv")

In [2]:
# Browsing the data imported to make sure it's in the right form
veg_df.head()

Unnamed: 0  840  1303  1313  1368  1362  1330  1304  1361  1363  ...  1212  \
0           1    0     3     3     0     3     3     3     3     3  ...     0   
1           2    1     0     3     3     0     0     3     0     0  ...     0   
2           3    3     3     3     3     3     3     3     3     3  ...     0   
3           4    0     0     3     3     0     0     3     0     0  ...     0   
4           5    0     0     0     0     0     0     0     0     0  ...     0   

   1211  1276  822  823  821  819  818  268  1244  
0     0     0    0    0    0    0    0    0     0  
1     0     0    0    0    0    0    0    0     0  
2     0     0    0    0    0    0    0    0     0  
3     0     0    0    0    0    0    0    0     0  
4     0     0    0    0    0    0    0    0     0  

[5 rows x 1422 columns]

In [3]:
# Removal of unwanted column and identifying the right index
veg_df = veg_df.set_index('Unnamed: 0')

In [4]:
# Browsing again the data imported after being positioned in the right form
veg_df.head()

840  1303  1313  1368  1362  1330  1304  1361  1363  1365  ...  \
Unnamed: 0                                                             ...   
1             0     3     3     0     3     3     3     3     3     3  ...   
2             1     0     3     3     0     0     3     0     0     3  ...   
3             3     3     3     3     3     3     3     3     3     3  ...   
4             0     0     3     3     0     0     3     0     0     0  ...   
5             0     0     0     0     0     0     0     0     0     0  ...   

            1212  1211  1276  822  823  821  819  818  268  1244  
Unnamed: 0                                                        
1              0     0     0    0    0    0    0    0    0     0  
2              0     0     0    0    0    0    0    0    0     0  
3              0     0     0    0    0    0    0    0    0     0  
4              0     0     0    0    0    0    0    0    0     0  
5              0     0     0    0    0    0    0    0    0     0  

[5 rows x 1421 columns]

In [5]:
# Importing the necessary libraries to perform Hopkins test analysis

from sklearn.neighbors import NearestNeighbors
from random import sample
from numpy.random import uniform
import numpy as np
from math import isnan
 
# Defining the Hopkins test function    
def hopkins(X):
    d = X.shape[1]
    #d = len(vars) # columns
    n = len(X) # rows
    m = int(0.1 * n) # heuristic from article [1]
    nbrs = NearestNeighbors(n_neighbors=1).fit(X.values)
 
    rand_X = sample(range(0, n, 1), m)
 
    ujd = []
    wjd = []
    for j in range(0, m):
        u_dist, _ = nbrs.kneighbors(uniform(np.amin(X,axis=0),np.amax(X,axis=0),d).reshape(1, -1), 2, return_distance=True)
        ujd.append(u_dist[0][1])
        w_dist, _ = nbrs.kneighbors(X.iloc[rand_X[j]].values.reshape(1, -1), 2, return_distance=True)
        wjd.append(w_dist[0][1])
 
    H = sum(ujd) / (sum(ujd) + sum(wjd))
    if isnan(H):
        print (ujd, wjd)
        H = 0
 
    return H

In [6]:
# Applying Hopkins test function on the imported data set
hopkins(veg_df)

0.9607421962080679

#### As shown above, the hypothesis test value is 0.96 which is quite close to 1, so we can reject the null hypothesis and this means that the data distribution is random implying the presence of clusters.